internal/indexer/oci.go

```diff
@@ -158,33 +159,34 @@ func (ind *ociIndexer) downloadIndex(ctx context.Context, rootPath string) (f st
 	ctx = remotes.WithMediaTypeKeyPrefix(ctx, chartsIndexLayerMediaType, "layer-")
 	ctx = remotes.WithMediaTypeKeyPrefix(ctx, chartsIndexConfigMediaType, "config-")
 
-	opts := []oras.PullOpt{
+	// Infer index filename from layer annotations
+	var indexFilename string
+	opts := []oras.CopyOpt{
 		oras.WithAllowedMediaType(chartsIndexLayerMediaType, chartsIndexConfigMediaType),
 		// The index artifact has no title
 		oras.WithPullEmptyNameAllowed(),
+		oras.WithLayerDescriptors(func(layers []ocispec.Descriptor) {
+			for _, layer := range layers {
+				switch layer.MediaType {
+				case chartsIndexLayerMediaType:
+					indexFilename = layer.Annotations["org.opencontainers.image.title"]
+				}
+			}
+		}),
 	}
-	_, layers, err := oras.Pull(ctx, ind.resolver, ind.reference, store, opts...)
+	// Fallback to the default index filename if the layers don't specify it
+	if indexFilename == "" {
+		klog.Infof("Unable to find index filename: using default")
+		indexFilename = defaultIndexFilename
+	}
+
+	_, err := oras.Copy(ctx, ind.resolver, ind.reference, store, ind.reference, opts...)
```

Comment:
Isn't this going to provoke a race condition on `indexFilename`?

Reply:
yep, you are right! This code blocks order is not correct. We should do the oras.Copy() before checking if indexFilename is empty